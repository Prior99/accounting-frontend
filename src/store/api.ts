import { store } from ".";

declare var baseUrl: string;

type HTTPMethod = "DELETE" | "GET" | "PUT" | "POST";

export async function api(url: string, body?: any, method: HTTPMethod = "GET"): Promise<any> {
    const headers = new Headers();
    if (store.login.loggedIn) {
        headers.append("authorization", store.login.authToken);
    }
    headers.append("content-type", "application/json");
    const fullUrl = `${window.location.protocol}//${baseUrl}${url}`;
    try {
        const response = await fetch(fullUrl, {
            method,
            headers,
            body: JSON.stringify(body),
        });
        const jsonResponse = (await response.json());
        const okay = response.status >= 200 && response.status <= 299;
        if (!okay) {
            store.error.apiError(`Fetching from url "${fullUrl}" resulted in error: "${jsonResponse.message}"`);
            return { okay };
        }
        return {
            okay,
            data: jsonResponse.data,
        };
    } catch (err) {
        store.error.apiError(`Failed to fetch from url "${fullUrl}".`, err);
        return { okay: false };
    }
}
