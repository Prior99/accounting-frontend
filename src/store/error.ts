import { observable, computed, action } from "mobx";

export interface DisplayableError {
    dismissed: boolean;
    message: string;
    origin: "api" | "internal";
    error?: Error;
}

export class ErrorStore {
    constructor() {
        this.apiError = this.apiError.bind(this);
        this.internalError = this.internalError.bind(this);
        this.dismissCurrent = this.dismissCurrent.bind(this);
    }

    @observable public errors: DisplayableError[] = [];

    @computed public get current() {
        return this.errors.find(error => !error.dismissed);
    }

    @action
    public dismissCurrent() {
        if (typeof this.current === "undefined") {
            return;
        }
        this.current.dismissed = true;
    }

    @action
    public apiError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "api", error });
    }

    @action
    public internalError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "internal", error });
    }
}
