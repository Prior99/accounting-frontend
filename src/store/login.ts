import { observable, computed, action } from "mobx";
import { api } from "./api";
import { bind } from "bind-decorator";

export class LoginStore {
    @observable public authToken: string;
    @observable public failed = false;

    @computed
    public get loggedIn() {
        return typeof this.authToken !== "undefined";
    }

    @bind
    @action
    public async onLogin(email: string, password: string) {
        const result = await api("/auth/login", { email, password }, "POST", true);
        const { data } = result;
        if (result.okay && data.auth_token) {
            this.authToken = data.auth_token;
        } else {
            this.failed = true;
        }
    }
}
