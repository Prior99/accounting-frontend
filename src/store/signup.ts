import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import { ApiStore } from ".";
import { History } from "history";
import { Store } from "./store";

export class SignupStore extends Store {
    @observable
    public signupResult: Boolean;

    @bind @action
    public async doSignup(email: string, password: string) {
        const body = { email, password };
        const response = await this.api.call("doSignup", () => this.api.registration.v1AuthRegisterPost({ body }));
        if (response) {
            await this.api.doLogin(email, password);
            this.signupResult = true;
        } else {
            this.signupResult = false;
        }
    }
}
