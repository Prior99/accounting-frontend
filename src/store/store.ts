import { ApiStore } from "./api";
import { History } from "history";

export class Store {
    protected api: ApiStore;
    protected browserHistory: History;

    constructor(api: ApiStore, browserHistory: History) {
        this.api = api;
        this.browserHistory = browserHistory;
    }
}
