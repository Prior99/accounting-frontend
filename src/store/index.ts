import { LoginStore } from "./login";
import { SignupStore } from "./signup";
import { ErrorStore } from "./error";
export * from "./error";

export interface Store {
    login: LoginStore;
    error: ErrorStore;
    signup: SignupStore;
}

export let store: Store;

export function createStore(): Store {
    return {
        login: new LoginStore(),
        error: new ErrorStore(),
        signup: new SignupStore()
    };
}

export function resetStore() {
    store = createStore();
}

resetStore();

if (typeof window !== "undefined") {
    (window as any).store = store;
    (window as any).createStore = createStore;
    (window as any).resetStore = resetStore;
}
