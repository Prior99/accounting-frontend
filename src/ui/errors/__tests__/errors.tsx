import * as React from "react";
import { mapStoreToProps, StrippedErrors } from "..";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { DisplayableError, store, resetStore } from "store";

const someError: DisplayableError = { dismissed: false, message: "some message", origin: "api" };

beforeEach(() => {
    resetStore();
});

test("The `Errors` component is not rendered if no error is currently active", () => {
    const mounted = mount(<StrippedErrors onDismiss={jest.fn()} />);
    expect(toJson(mounted)).toMatchSnapshot();
});

test("The `Errors` component renders the current error", () => {
    const mounted = mount(<StrippedErrors error={someError} onDismiss={jest.fn()} />);
    expect(toJson(mounted)).toMatchSnapshot();
});

test("`mapStoreToProps` infers an `DisplayableError` if present", () => {
    store.error.apiError("some message");
    const result = mapStoreToProps(store);
    expect(result).toMatchSnapshot();
});
