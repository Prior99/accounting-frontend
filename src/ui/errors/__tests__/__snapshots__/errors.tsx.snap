// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Errors\` is rendered correctly with a network error 1`] = `
<Errors>
  <Modal
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    onClose={[Function]}
    open={true}
  >
    <Portal
      className="ui page modals dimmer transition visible active"
      closeOnDocumentClick={false}
      closeOnEscape={true}
      closeOnRootNodeClick={true}
      mountNode={
        <body
          class="dimmable dimmed"
        >
          <div
            class="ui page modals dimmer transition visible active"
          >
            <div
              data-reactroot=""
              class="ui modal transition visible active"
            >
              <div
                class="header"
              >
                An error occured
              </div>
              <div
                class="content"
              >
                <div
                  class="description"
                >
                  some.message
                </div>
              </div>
            </div>
          </div>
        </body>
      }
      onClose={[Function]}
      onMount={[Function]}
      onOpen={[Function]}
      onUnmount={[Function]}
      open={true}
      openOnTriggerClick={true}
    />
  </Modal>
</Errors>
`;

exports[`\`Errors\` is renders nothing without an error 1`] = `<Errors />`;
